### DISCUSSION TITLE
Is there a way to catch plt.show for frontends?

### DISCUSSION THREAD

**Discussion by ngmisl** [sinaptik-ai/pandas-ai#166] [Q&A]
I'm trying to catch the plot via

```
        if isinstance(output, Image.Image):  # if the output is an image
            buffered = io.BytesIO()
            output.save(buffered, format="JPEG")
            img_str = base64.b64encode(buffered.getvalue())
            return JSONResponse(
                {"type": "image", "data": img_str.decode("utf-8")}
```

and displaying it via javascript on the front-end. But it seems pandas-ai does not pass the relevant data?

Is there a way to modify this?

**Comment by gventuri** [comment]
@ngmisl where do you run this code?
Guess as of now we don't have such an option, but it would definitely be a nice addition.
An alternative would be to save it (we are working on it) on the server and return the url.

**Reply by ngmisl** [reply]
@gventuri in a fastapi. But I can't store data on the server in production. It works if you add a negative prompt (don't use plt.show()) but that's saving the output then. I just need the base64 passed somehow, would open up any use case then

**Comment by mspronesti** [comment]
@ngmisl you can do that asking pandasai to that all of this for you in the prompt. For instance, adapting `from_chart.py` 
```text
    Generate the histogram of countries showing for each the gpd, using different colors for each bar, encode it to base64 and output the result in json format.
    Json: "
```

The only problem here would be that in the latest weeks pandas-ai has been downpowered and the generated code wouldn't be executed because it'd contain `json`, `base64` and `io` which are not whitelisted. If you add these to the dictionaries of allowed libs and builtins, the generated code would be something like this

```py
import pandas as pd
import matplotlib.pyplot as plt
import base64
import io
import json

# create dataframe
data = {'country': ['United Kingdom', 'United States', 'Germany', 'Italy', 'France'],
        'gdp': [2891615567872, 8469645243, 444827455, 6820579572, 5381911486],
        'happiness_index': [6.66, 6.38, 7.07, 7.16, 6.38]}
df = pd.DataFrame(data)

# plot histogram
fig, ax = plt.subplots()
colors = ['r', 'g', 'b', 'c', 'm']
for i in range(len(df)):
    ax.bar(df['country'][i], df['gdp'][i], color=colors[i])
ax.set_xlabel('Country')
ax.set_ylabel('GDP')

# encode to base64 and output in json format
buffer = io.BytesIO()
plt.savefig(buffer, format='png')
buffer.seek(0)
image_png = buffer.getvalue()
buffer.close()
graph = base64.b64encode(image_png).decode('utf-8')
result = {'graph': graph}
json_output = json.dumps(result)
print(json_output)
```

Hope this helps.
@gventuri Do you think we should take a step back on this imports restriction ?

**Reply by gventuri** [reply]
@mspronesti doing a step back sounds a little bit scaring to me, but on the other hand I agree we should add more libraries to the whitelist and fix a couple of known issues with the current implementation.

From what we've seen in other cases, the imports seem to be ignored no matter what, and they only work if we pass them as builtins in the exec/eval!

**Reply by mspronesti** [reply]
I will probably add an issue to discuss this much more into detail :)

**Reply by ngmisl** [reply]
Yeah could be done via prompting I guess. But that's not feasible for the end user I have in mind. They don't know about prompting, and the NLP aspect is the biggest selling point. Hope you can implement those features, I would really like to use pandasai in a little consumer app

**Reply by gventuri** [reply]
@ngmisl it should work fine now with the version 0.5.1. Let me know!!
